---
title: "methylSig: A package for whole genome DNA methylation analysis"
author: "Yongseok Park, Raymond G. Cavalcante, Maria E. Figueroa, Laura S. Rozek, and Maureen A. Sartor"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to methylSig}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE}
library(methylSig)
library(rtracklayer)
```

# Introduction

DNA methylation plays critical roles in gene regulation and cellular specification without altering DNA sequences. It is one of the best understood and most intensively studied epigenetic marks in mammalian cells. Treatment of DNA with sodium bisulfite deaminates unmethylated cytosines to uracil while methylated cytosines are resistant to this conversion thus allowing for the discrimination between methylated and unmethylated CpG sites. Sodium bisulfite pre-treatment of DNA coupled with next-generation sequencing has allowed DNA methylation to be studied quantitatively and genome-wide at single cytosine site resolution.

`methylSig` is a method for testing for differential methylated cytosines (DMCs) or regions (DMRs) in whole-genome bisulfite sequencing (bis-seq) or reduced representation bisulfite sequencing (RRBS) experiments. `methylSig` uses a beta binomial model to test for significant differences between groups of samples. Several options exist for either site-specific or sliding window tests, combining strands, and for variance estimation. It allows annotating the resulting regions to multiple genome features, and visualizing the results for chosen genomic regions.

# Installation

`methylSig` is available on GitHub at <http://www.github.com/sartorlab/methylSig>. Please use the appropriate branch for your version of R / Bioconductor:

## R 3.4 and Bioconductor 3.6

```{r, eval=FALSE}
devtools::install_github('sartorlab/methylSig@R3.4_Bioc3.6')
```

## R 3.5 and Bioconductor 3.7

Forthcoming.

## R 3.5 and Bioconductor 3.8

```{r, eval=FALSE}
devtools::install_github('sartorlab/methylSig@R3.5_Bioc3.8')
```

# Basic usage

## Reading data

As of version 0.5.0, `methylSig` is able to read `bismark_methylation_extractor` outputs directly using the `bsseq` Bioconductor package. The `methylSigReadData()` function is a wrapper for `bsseq::read_bismark()` that adds some userful features:

1. Users can set a `minCount` and `maxCount` for the coverage of sites.
2. Users analyzing data aligned to `hg19` can filter out C > T or G > A SNPs.

The following code uses data contained in the package to demonstrate how to read methylation data:

```{r}
# The following bismark cytosine reports are included in inst/extdata
files = c(
    system.file('extdata', 'MDAMB_231_1DR.txt.gz', package='methylSig'),
    system.file('extdata', 'MDAMB_231_1DS.txt.gz', package='methylSig'),
    system.file('extdata', 'MDAMB_231_2DR.txt.gz', package='methylSig'),
    system.file('extdata', 'MDAMB_231_2DS.txt.gz', package='methylSig'),
    system.file('extdata', 'MDAMB_231_3DR.txt.gz', package='methylSig'),
    system.file('extdata', 'MDAMB_231_3DS.txt.gz', package='methylSig'))

sample.ids = basename(files)
sample.ids = gsub('.txt.gz', '', sample.ids)

# Build a pData matrix with columns for the samples, group memberships, and phenotype data
pData = data.frame(
    Sample_Names = sample.ids,
    DR_vs_DS = relevel(factor(c('DR','DS','DR','DS','DR','DS')), ref = 'DS'),
    row.names = sample.ids,
    stringsAsFactors = FALSE)

meth = methylSigReadData(
    fileList = files,
    pData = pData,
    assembly = 'hg19',
    destranded = TRUE,
    maxCount = 500,
    minCount = 10,
    filterSNPs = TRUE,
    num.cores = 1)

print(meth)
```

## Differential methylation analysis

The package consists of two methods to test for differential methlyation: `methylSigCalc()` and `methylSigDSS()`.

The first, `methylSigCalc()`, calculates differential methylation statistics between two groups of samples. It uses a beta-binomial approach to calculate differential methylation statistics, accounting for coverage and variation among samples within each group.

The second, `methylSigDSS()`, is a wrapper for the `DSS::DMLfit.multiFactor()` and `DSS::DMLtest.multiFactor()` functions in the [DSS Bioconductor package](https://bioconductor.org/packages/release/bioc/html/DSS.html). Essentially the test in DSS uses a linear model over an arbitrary design matrix, thus allowing for correction by covariates. The wrapper function provided here enables enforcement of the a minimum number of data points per group to test a site/region, as well as reporting of averaged methlyation levels over the groups.

### Site specific analysis with `methylSigCalc()`

The default is to do site specific analysis and to use both groups to estimate variances.

```{r}
### Test on CpGs
result = methylSigCalc(
    meth = meth,
    comparison = 'DR_vs_DS',
    dispersion = 'both',
    local.info = FALSE,
    local.winsize = 200,
    min.per.group = c(3,3),
    weightFunc = methylSig_weightFunc,
    T.approx = TRUE,
    num.cores = 1)

print(result)
```

The output includes the estimated dispersion (`phiCommonEst`), the log-likelihood ratio (`logLikRatio`), the degrees of freedom (`df`), the group methylation estimates (`muEstC_group1` and `muEstC_group2` where group 1 is the reference factor in the comparison column of the `pData` matrix), the methylation difference (`meth.diff = muEstC_group2 - muEstC_group1`), the group for which the site is hyper-methylated (`hyper.direction`, note, this is regardless of significance), the `pvalue`, and `fdr`.

#### Variance from one group

Using the `dispersion` argument, it is possible to estimate variances from one group rather than from both groups. This can be accomplished by changing the `dispersion` parameter in the previous example from `'both'` to `'DS'` or `'DR'`

#### Using local information

It is also possible to use information from nearby CpG sites to improve the variance and methylation level estimates. The default `local.winsize` is 200 bps.  The `local.winsize` is only used when `local.info = TRUE`.

### Site specific analysis with `methylSigDSS()`

The following example illustrates a case with no covariates and where the model has an intercept.

```{r}
# Must create a design matrix
design1 = data.frame(group = bsseq::pData(meth)$DR_vs_DS)

print(design1)

# NOTE this model has an intercept
contrast_intercept = matrix(c(0,1), ncol = 1)
result_dss_intercept = methylSigDSS(
    meth = meth,
    design = design1,
    formula = '~ group',
    contrast = contrast_intercept,
    group.term = 'group',
    min.per.group=c(3,3))

print(result_dss_intercept)
```

The following illustrates an paired-type test.

```{r}
# Add a covariate column, note specification as a factor, but can
# also use a numeric covariate
design2 = data.frame(
    group = bsseq::pData(meth)$DR_vs_DS,
    subject = factor(c(1,1,2,2,3,3)))

print(design2)

# NOTE the contrast vector has as many entries as the sum of the
# levels in group and subject, in the formula.
contrast_covariates = matrix(c(0,1,0,0), ncol = 1)
result_dss_covariates = methylSigDSS(
    meth = meth,
    design = design2,
    formula = '~ group + subject',
    contrast = contrast_covariates,
    group.term = 'group',
    min.per.group=c(3,3))

print(result_dss_covariates)
```

### Tiled analysis

`methylSig` also provides `methylSigTile()` to tile data within continuous non-overlapping windows. Users can tile the genome according to a window size, give a `data.frame` with genomic regions, or give a `GRanges` object. Examples are below. Note that tiling analysis is also possible with `methylSigDSS()`.

#### Windowed analysis

```{r}
### Test on 10000bp windows
windowed_meth = methylSigTile(meth, tiles = NULL, win.size = 10000)

tiled_result = methylSigCalc(
    meth = windowed_meth,
    comparison = 'DR_vs_DS',
    dispersion = 'both',
    local.info = FALSE,
    local.winsize = 200,
    min.per.group = c(3,3),
    weightFunc = methylSig_weightFunc,
    T.approx = TRUE,
    num.cores = 1)

print(tiled_result)
```

#### Region analysis

As mentioned, users can provide a `data.frame` in the `tiles` parameter, so long as it has column names acceptable to `makeGRangesFromDataFrame`, i.e. `chr`, `start`, and `end`.

Finally, users can provide tiling regions as a `GRanges` object. If we wanted to test for differential methylation in CpG islands, we could use the `annotatr` package to create the CpG island regions.

```{r}
### Test on CpG islands
cpg_islands = annotatr::build_annotations(genome = 'hg19', annotations = 'hg19_cpg_islands')

cpg_island_meth = methylSigTile(meth, tiles = cpg_islands)

cpg_island_result = methylSigCalc(
    meth = cpg_island_meth,
    comparison = 'DR_vs_DS',
    dispersion = 'both',
    local.info = FALSE,
    local.winsize = 200,
    min.per.group = c(3,3),
    weightFunc = methylSig_weightFunc,
    T.approx = TRUE,
    num.cores = 1)

print(cpg_island_result)
```

# Annotation

## Annotating differentially methylated CpGs

Once differential methylation has been determined with `methylSigCalc()`, it may be of interest understand where differential methylation occurs in terms of genes and CpG features (islands, shores, shelves). `methylSig` uses the `annotatr` package to accomplish this.

```{r}
# Get CpG island annotations from built-in data they could be built with the following:
# cpg_annots = annotatr::build_annotations(genome = 'hg19', annotations = c('hg19_cpg_islands'))
utils::data(sample_data, package = 'methylSig')

# Determine what CpGs should be considered significant
dmcList = result$fdr < 0.05 & abs(result$meth.diff) > 25

annotated_result = methylSigAnnotation(myDiff = result, dmcList = dmcList, annotations = cpg_annots)
```

The result is a `GRanges` object with the same columns as the `result` with the addition of columns giving the differential methylation status (`dm_status`), a unique locus id (`locus_id`), and information about the annotation which is itself a `GRanges` object (`annot`). It is important to note that regions tested for differential methylation may occur on multiple rows, depending on the number of features it is annotated to. The `locus_id` column helps to quickly see when this is the case.

```{r}
print(annotated_result)
```

It is more illuminating to view this object as a coerced `data.frame`, wherein the information about the annotations are displayed.

```{r}
print(head(as.data.frame(annotated_result)))
```

## Transcription factor (TF) enrichment test

Changes in DNA methylation have been shown to alter transcription factor binding. The `methylSig` package has implemented `methylSig.tfbsEnrichTest()` to test a set of transcription factor binding sites (TFBSs) are enriched for differentially methylated CpGs.

We demonstrate this funciton on a set of TFBSs from ENCODE. The `rtracklayer::import()` function makes reading in `BED` files simple.

```{r}
# Use preloaded tfbs from package sample_data. Could be manually loaded as with:
# tfbs_file = system.file('extdata','tfbs.bed.gz', package = 'methylSig')
# tfbs = rtracklayer::import(tfbs_file, genome = 'hg19')

print(tfbs)
```

This file mixes TFBSs from a number of TFs and keeps track of them in the name column (4th) of the `BED`. Next, we indicate what is considered a differentially methylated CpG and perform the test.

```{r}
# Significance threshold
dmcList = result$fdr < 0.05 & abs(result$meth.diff) > 25

# Perform the test
tfbs_enrichment = methylSig.tfbsEnrichTest(myDiff = result, dmcList = dmcList, tfbsInfo = tfbs)

# Take a look at the first few rows
print(head(tfbs_enrichment))
```
